platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ImportMigrationConfigurationAdminContext.java

```diff
@@ -0,0 +1,234 @@
+import java.nio.file.PathMatcher;
+import java.util.Dictionary;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.io.FilenameUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.Validate;
+import org.codice.ddf.configuration.persistence.PersistenceStrategy;
+import org.codice.ddf.migration.ImportMigrationContext;
+import org.codice.ddf.migration.ImportMigrationContextProxy;
+import org.codice.ddf.migration.ImportMigrationEntry;
+import org.codice.ddf.migration.MigrationException;
+import org.codice.ddf.migration.MigrationReport;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class extends on the {@link ImportMigrationContext} interface to pre-process exported entries
+ * for configuration objects and compare them with the configuration objects currently in memory.
+ */
+public class ImportMigrationConfigurationAdminContext extends ImportMigrationContextProxy {
+    private static final Logger LOGGER = LoggerFactory.getLogger(
+            ImportMigrationConfigurationAdminContext.class);
+
+    private final ConfigurationAdminMigratable admin;
+
+    private final ConfigurationAdmin configurationAdmin;
+
+    /**
+     * Keeps track of all managed services in memory that were not found in the export file so
+     * we know what to delete at the end.
+     */
+    private final Map<String, Configuration> memoryServices;
+
+    /**
+     * Keeps track of all managed service factories in memory that were not found in the export file
+     * so we know what to delete at the end.
+     */
+    private final Map<String, List<Configuration>> memoryFactoryServices;
+
+    private final Map<String, ImportMigrationConfigurationAdminEntry> exportedServices;
+
+    private final Map<String, List<ImportMigrationConfigurationAdminEntry>> exportedFactoryServices;
+
+    private boolean isValid = true; // until proven otherwise
+
+    public ImportMigrationConfigurationAdminContext(ImportMigrationContext context,
+            ConfigurationAdminMigratable admin, ConfigurationAdmin configurationAdmin,
+            Configuration[] memoryConfigs) {
+        super(context);
+        Validate.notNull(admin, "invalid null configuration admin migratable");
+        Validate.notNull(configurationAdmin, "invalid null configuration admin");
+        Validate.notNull(memoryConfigs, "invalid null configurations");
+        this.admin = admin;
+        this.configurationAdmin = configurationAdmin;
+        // categorize memory configurations
+        this.memoryServices = Stream.of(memoryConfigs)
+                .filter(ConfigurationAdminMigratable::isManagedService)
+                .collect(Collectors.toMap(Configuration::getPid, Function.identity()));
```

Comment:
❓ Do we need to worry about `getPid()` returning `null`?

Reply:
I don't think so. As far as I can tell, a config object cannot exist without a PID.